[#41] Firebase Cloud Messaging을 활용한 Push 알람 구현

msugo1 opened the pull request:
## 📖 Model 관련

### 📑  PushNotificationRequest

- 푸시 알람과 관련한 요청에 사용할 인스턴스를 만들 클래스 생성

- 푸시 알람을 보내는 targetId, 제목, 그리고 메시지를 멤버 변수로 가짐

### 📑 UserLoginRequest

- 토큰도 함께 서버로 전달될 수 있도록 `token` 변수 추가

- Builder를 추가 해 token 없이도 테스트 시 인스턴스를 생성할 수 있게 변경

### 📑 UserLoginData

- 사용자가 로그인 요청을 위해 보낸 ID와 일치하는 로그인 정보를 DB에서 받아올 객체를 생성하기 위한 클래스 생성

## 📖 Enum 관련

### 📑 AlarmTypes

- 푸시 알람에 사용할 수 있는 message 값 또한 멤버 변수로 추가

## 📖 Service 계층 관련

### 📑 PushNotificationService, FirebasePushNotificationService

- PushNotification은 FCM 외에도 다른 솔루션을 사용할 수 있으므로, 기본 뼈대인 인터페이스를 두고 이를 구현하도록 설계

- FCM 서비스를 활성화하는 init 메소드 추가

- 사용자가 로그인 시 보낸 token을 Redis 저장소에 등록하는 `registerToken` 메소드 추가

- 저장한 토큰을 얻어오는 `getToken` 메소드 추가

- 메시지를 만들고 푸시 알람을 만들 `sendPushNotification` 메소드 추가

### 📑 FriendService, RecommendationService

 - 친구 추가 요청, 친구 추가 요청 승낙, 추천글 등록 시 푸시 알람을 함께 보낼 수 있도록 코드 추가

### 📑 SessionLoginService

- 중복 로그인 검사 코드를 별도의 메소드로 추출

- 로그인 시 DB에서 유저 정보를 모두 받아오는 것이 아니라 필요한 정보만 받아오도록 변경 (id, password, 조인 후 token)

- 받아온 토큰은 로그인 시 함께 들어온 토큰과 일치여부 검사 후 일치하지 않은 경우만 DB에 업데이트 반영

### 📑 UserService

- `getUser -> getUserLoginInfo`로 메소드 이름 변경

- 로그인에 사용할 사용자 정보를 전체 사용자 정보가 아닌 필요한 정보만
  리턴하도록 변경

## 📖 Mapper 계층 관련

### 📑 UserMapper, UserMapper.xml

- getUserLoginInfo로 메소드를 변경한 부분 업데이트 반영

- 조인 후 사용자에게 저장된 토큰을 받아올 수 있도록 쿼리 변경

### 📑 PushNotificationMapper, PushNotificationMapper.xml

- DB에 토큰을 저장/업데이트하고 받아올 `upsertToken`, `getToken` 메소드 추가

- 위의 내용을 처리할 쿼리 추가 (특히, upsertToken의 경우 INSERT INTO ~  ON DUPLICATE KEY UPDATE 쿼리 사용)

## 📖 테스트 관련

- 위의 변경 사항을 테스트 코드에 반영

- 테스트 코드의 동작 확인

Commit:
[#41] Firebase Cloud Messaging을 사용해 Push 알람 구현

Model 관련

* PushNotificationRequest

- 푸시 알람과 관련한 요청에 사용할 인스턴스를 만들 클래스 생성

- 푸시 알람을 보내는 targetId, 제목, 그리고 메시지를 멤버 변수로 가짐

* UserLoginRequest

- 토큰도 함께 서버로 전달될 수 있도록 `token` 변수 추가

- Builder를 추가 해 token 없이도 테스트 시 인스턴스를 생성할 수 있게
  변경

Enum 관련

* AlarmTypes

- 푸시 알람에 사용할 수 있는 message 값 또한 멤버 변수로 추가

Controller 계층 관련

* UserController

- login 요청을 받을 때 클라이언트 토큰이 함께 들어온다면 이를
  세션과 함께 저장해두도록 설정

- logout 요청을 받을 때 클라이언트 토큰도 함께 삭제

Service 계층 관련

* PushNotificationService, FirebasePushNotificationService

- PushNotification은 FCM 외에도 다른 솔루션을 사용할 수 있으므로,
  기본 뼈대인 인터페이스를 두고 이를 구현하게 함으로써 OCP를 달성하고자
  함

- FCM 서비스를 활성화하는 init 메소드 추가

- 사용자가 로그인 시 보낸 token을 Redis 저장소에 등록하는 `registerToken` 메소드 추가

- 저장한 토큰을 얻어오는 `getToken` 메소드 추가

- 사용자가 로그인 시 저장된 토큰을 삭제할 `destroyToken` 메소드 추가

- 메시지를 만들고 푸시 알람을 만들 `sendPushNotification` 메소드 추가

* FriendService, RecommendationService

- 친구 추가 요청, 친구 추가 요청 승낙, 추천글 등록 시 푸시 알람을 함께
  보낼 수 있도록 코드 추가

Commit:
[#41] 토큰 저장 구현 변경

Service 계층 관련

* SessionLoginService

- 중복 로그인 검사 코드를 별도의 메소드로 추출

- 로그인 시 DB에서 유저 정보를 모두 받아오는 것이 아니라 필요한 정보만
  받아오도록 변경 (id, password, 조인 후 token)

- 받아온 토큰은 로그인 시 함께 들어온 토큰과 일치여부 검사 후 일치하지
  않은 경우만 DB에 업데이트 반영

* UserService

- `getUser -> getUserLoginInfo`로 메소드 이름 변경

- 로그인에 사용할 사용자 정보를 전체 사용자 정보가 아닌 필요한 정보만
  리턴하도록 변경

* FirebasePushNotificationService

- 토큰을 등록할 때와 토큰을 얻어올 때, 저장소를 RDBMS로 변경

Mapper 계층 관련

* UserMapper, UserMapper.xml

- getUserLoginInfo로 메소드를 변경한 부분 업데이트 반영

- 조인 후 사용자에게 저장된 토큰을 받아올 수 있도록 쿼리 변경

* PushNotificationMapper, PushNotificationMapper.xml

- DB에 토큰을 저장/업데이트하고 받아올 `upsertToken`, `getToken` 메소드 추가

- 위의 내용을 처리할 쿼리 추가 (특히, upsertToken의 경우 INSERT INTO ~
  ON DUPLICATE KEY UPDATE 쿼리 사용)

Model 관련

* UserLoginData

- 사용자가 로그인 요청을 위해 보낸 ID와 일치하는 로그인 정보를 DB에서 받아올 객체를 생성하기 위한 클래스 생성

테스트 관련

- 위의 변경 사항을 테스트 코드에 반영

- 테스트 코드의 동작 확인

msugo1 commented:
Can one of the admins verify this patch?

Commit:
[#41] 불필요한 의존성 주입 제거

- `Redis -> MySQL`로 토큰 저장소를 바꿀 때 제거하지 않았던
  RedisTemplate에 대한 의존성을 제거

- RedisTemplate을 사용하면서 Test Scope에서 Configuration 문제가
  해결되었으므로 RedisSessionConfig의 프로필을 다시 `prod`, `local`로
  설정

Commit:
통합 테스트 클래스 이름 변경

f-lab-dev commented on src/main/java/me/soo/helloworld/service/FirebasePushNotificationService.java:
```diff
@@ -0,0 +1,89 @@
+        Message pushMessage = writePushMessage(pushNotificationRequest, token);
+        ApiFuture<String> response = FirebaseMessaging.getInstance().sendAsync(pushMessage);
+        try {
+            log.info("{} 님에게 푸시 알람 전송을 전송하였습니다. : {}", pushNotificationRequest.getTargetId(), response.get());
```
`response.get()`으로 blocking 시키기보다는 콜백을 등록해주는게 좋을 것 같네요~

msugo1 commented:
Hello!

Commit:
[#41] 피드백 반영

- 비동기 전송 후 `blocking`을 유발 해 비동기 전송의 장점을 무마시킨
  an ApiFuture instance.get 메소드 실행 부분 제거

- UserMapper.xml getUserLoginData 부분 resultMap 오류 Fix

Commit:
불필요한 import 라인 제거

f-lab-dev commented on src/main/java/me/soo/helloworld/service/FirebasePushNotificationService.java:
```diff
@@ -67,12 +71,7 @@ public void sendPushNotification(PushNotificationRequest pushNotificationRequest
-        } catch (InterruptedException | ExecutionException e) {
-            throw new IllegalStateException("푸시알람 전송에 실패하였습니다.", e);
-        }
+        FirebaseMessaging.getInstance().sendAsync(pushMessage);
```
실패했을 경우 로그 남기는건 해주면 좋지 않을까요?
